Add shift-click feature on memory view to put focus on pointed address

Vancleeff opened the pull request:
Enhanced user experience by providing a convenient way to navigate to specific pointed memory addresses

Vancleeff committed:
Add shift-click feature on memory view to put focus on pointed address

Enhanced user experience by providing a convenient way to navigate to specific pointed memory addresses

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -264,6 +268,15 @@ void MemoryViewerControlBinding::OnClick(POINT point)
+    const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+    ra::ByteAddress nAddress =
+        pEmulatorContext.ReadMemory(m_pViewModel.GetAddress(), m_pViewModel.GetSize()) & 0xFFFFFF;
+    m_pViewModel.OnShiftClick(nAddress);
```
You're not shift+clicking `nAddress`, you're just setting the address after converting it. The conversion should be here, then just call `SetAddress`.

Although, for unit testing purposes all of this code should be in the view model. Suggest making `MemoryViewerViewModel::OnShiftClick` accept an x and y, have that call `MemoryViewerViewModel::OnClick`, do the address calculations and call `MemoryViewerViewModel::SetAddress`. Then, please, write unit tests for `MemoryViewerViewModel::OnShiftClick`.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -34,7 +35,10 @@ INT_PTR CALLBACK MemoryViewerControlBinding::WndProc(HWND hControl, UINT uMsg, W
 
         case WM_LBUTTONUP:
-            OnClick({ GET_X_LPARAM(lParam) - MEMVIEW_MARGIN, GET_Y_LPARAM(lParam) - MEMVIEW_MARGIN });
+            if (bShiftHeld)
```
if `bShiftHeld` is only needed for this case, recommend only evaluating it here.
```
if (GetKeyState(VK_SHIFT) < 0)
```

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -264,6 +268,15 @@ void MemoryViewerControlBinding::OnClick(POINT point)
 
+void MemoryViewerControlBinding::OnShiftClick(POINT point)
+{
+    OnClick(point);
```
Note that `OnClick` will abort if the memory viewer has a fixed address (such as when it's used in the pointer finder). This functionality should also be non-functional in the pointer finder.

Jamiras commented on src/ui/win32/bindings/MemoryViewerControlBinding.cpp:
```diff
@@ -264,6 +268,15 @@ void MemoryViewerControlBinding::OnClick(POINT point)
+    OnClick(point);
+    const auto& pEmulatorContext = ra::services::ServiceLocator::Get<ra::data::context::EmulatorContext>();
+    ra::ByteAddress nAddress =
+        pEmulatorContext.ReadMemory(m_pViewModel.GetAddress(), m_pViewModel.GetSize()) & 0xFFFFFF;
```
Why are you masking the value? That should be implicit within `ByteAddressFromRealAddress`

Vancleeff committed:
Refactor pointer follower feature following review
 - Moved logic to ViewModel
 - Removed 'OnShiftClick' function in ControlBinding and simply add a keystate check on the 'OnClick' one
 - Removed unecessary mask
 - Unit Test added

vancleeff commented:
Thank you for the review @Jamiras, very helpful !
I made some correction and pushed everything, I'm not sure if the tests are enough for this usecase or if I need to add more console context.

Jamiras commented on tests/ui/viewmodels/MemoryViewerViewModel_Tests.cpp:
```diff
@@ -1918,8 +1922,78 @@ TEST_CLASS(MemoryViewerViewModel_Tests)
+        viewer.OnShiftClick(10 * CHAR_WIDTH, CHAR_HEIGHT + 4);
+        Assert::AreEqual({0x20U}, viewer.GetAddress());
+
+        // Shift click on the second word containing 0x40 should lead to address 0xFF
```
Why would 0x40 lead to 0xFF? I think the comment is wrong.

Also, I'd prefer if this second click actually went to a 16-bit address (like 0x0140). The 0xFFFF test case handles the extreme location, but it would be good to validate a 16-bit valid address. You can always call `SetAddress(0)` to reset things for the next click.

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.cpp:
```diff
@@ -895,6 +896,21 @@ void MemoryViewerViewModel::OnClick(int nX, int nY)
+    ra::ByteAddress nAddress = pEmulatorContext.ReadMemory(GetAddress(), GetSize());
+    const auto nConvertedAddress = pConsoleContext.ByteAddressFromRealAddress(nAddress);
+    if (nConvertedAddress != 0xFFFFFFFF)
+        nAddress = nConvertedAddress;
```
Please include a test for an inconvertible address.

Vancleeff committed:
Update following review
 - Improve unit tests
 - Add a mask to handle complex pointer like PSX

vancleeff commented:
Thanks @Jamiras ! I changed the logic a bit by adding a mask for more complex pointer which are not converted into real addresses when using `ByteAddressFromRealAddress`. 

Example : PSX pointer like `0x80012345` are now masked based on the `pConsoleContext.MaxAddress` size and then converted with `ByteAddressFromRealAddress` to point towards `0x12345` address.

I adapted the tests accordingly.

Jamiras commented on src/ui/viewmodels/MemoryViewerViewModel.cpp:
```diff
@@ -895,6 +896,31 @@ void MemoryViewerViewModel::OnClick(int nX, int nY)
+
+    auto nMask = std::stoi(sMask, 0, 16);
+    ra::ByteAddress nAddress = (pEmulatorContext.ReadMemory(GetAddress(), GetSize())) & nMask;
+    const auto nConvertedAddress = pConsoleContext.ByteAddressFromRealAddress(nAddress);
```
You completely misunderstood my last comment. What you had was fine. I just wanted a test added for clicking on an address that was not convertible.

You can't convert an address just by masking. It may work for PSX, but it wouldn't work for GBA. The real address "$02001234" converts to a RetroAchievements address of "$00009234".
```
{ 0x008000U, 0x047FFFU, 0x02000000U, RC_MEMORY_TYPE_SYSTEM_RAM, "System RAM" },
```

The conversion is: subtract 0x02000000, then add 0x008000.

The point of calling `ByteAddressFromRealAddress` is to let it do the conversion for you. Just provide the raw data and it'll return an RA Address, or 0xFFFFFFFF. I just wanted you to add a test for a case where it returned 0xFFFFFFFF so we could enshrine the intended behavior. The previous behavior would have just jumped to the max address. The current implementation will not jump at all, which I feel is better.

Jamiras replied:
I think I see why you're trying to mask now. The `ConsoleContext` is initialized from the rcheevos regions, and the PlayStation regions only define the real memory, not the mirrors:
```
static const rc_memory_region_t _rc_memory_regions_playstation[] = {
    { 0x000000U, 0x00FFFFU, 0x000000U, RC_MEMORY_TYPE_SYSTEM_RAM, "Kernel RAM" },
    { 0x010000U, 0x1FFFFFU, 0x010000U, RC_MEMORY_TYPE_SYSTEM_RAM, "System RAM" }
};
```
There are mirrors at $80000000 and $A0000000 that should also map to $00000000 (real address), which translates to $00000000 (RetroAchievements address).

This logic should be implemented in `ByteAddressFromRealAddress` on a per-console basis. Generic masking is not the correct solution.

Vancleeff replied:
Ok, now I understand how it works, it was a little confusing because of the PSX use-case.
I added these 2 mirrored RAM to consoleinfo.c in rcheevos and now, it works perfectly well on PSX too !
Thanks for your patience @Jamiras !

Vancleeff replied:
To be sure I did it correctly, here's a commit with the change I made to consoleinfo.c : https://github.com/RetroAchievements/rcheevos/commit/59cd2d2a25fa52d57071c1e42a421108c5907366

Jamiras replied:
No, no, no. The regions defined in rcheevos tell emulator developers what memory we expect to be able to find. If you add these mirror regions to the memory map, then they'll become available for writing achievements and appear in search results. See [mirror RAM in NES](https://docs.retroachievements.org/Console-Specific-Tips/#memory-mirroring) and [echo RAM in GameBoy](https://docs.retroachievements.org/Console-Specific-Tips/#memory-mirroring). Not all cores expose the duplicated memory, so achievements relying on it sometimes don't work.

What I meant was something more like this:
```
ra::ByteAddress ConsoleContext::ByteAddressFromRealAddress(ra::ByteAddress nRealAddress) const noexcept
{
    // Map mirrored memory addresses to non-mirrored addresses.
    // Usually, only the non-mirrored addresses appear in m_vRegions.
    switch (m_nId)
    {
        case RC_CONSOLE_PLAYSTATION:
            if ((nRealAddress & 0xFFE00000) == 0x80000000 || // Kernel and User Memory Mirror (2 Meg) Cached
                (nRealAddress & 0xFFE00000) == 0xA0000000)   // Kernel and User Memory Mirror (2 Meg) Uncached
            {
                nRealAddress &= 0x001FFFFF;
            }
            break;

        case RC_CONSOLE_PLAYSTATION_2:
            if ((nRealAddress & 0xFE000000) == 0x20000000 || // Main RAM, uncached
                (nRealAddress & 0xFE000000) == 0x30000000)   // Main RAM, uncached and accelerated
            {
                nRealAddress &= 0x01FFFFFF;
            }
            break;
    }

    for (const auto& pRegion : m_vRegions)
    {
        if (pRegion.RealAddress < nRealAddress)
        {
            const auto nRegionSize = pRegion.EndAddress - pRegion.StartAddress;
            if (nRealAddress < pRegion.RealAddress + nRegionSize)
                return (nRealAddress - pRegion.RealAddress) + pRegion.StartAddress;
        }
    }

    return 0xFFFFFFFF;
}

```

Jamiras commented on tests/ui/viewmodels/MemoryViewerViewModel_Tests.cpp:
```diff
@@ -1918,8 +1922,120 @@ TEST_CLASS(MemoryViewerViewModel_Tests)
+        ra::data::context::mocks::MockConsoleContext mockConsole(PlayStation, L"Playstation");
+
+        MemoryViewerViewModelHarness viewer;
+        viewer.InitializeMemory(0x1FFFFF); // PSX full memory
```
Do you really need to allocate all 16MB?

Jamiras commented on tests/ui/viewmodels/MemoryViewerViewModel_Tests.cpp:
```diff
@@ -1918,8 +1922,120 @@ TEST_CLASS(MemoryViewerViewModel_Tests)
+        // Shift click on the first dword containing 0x80012345 should lead to address 0x00012345 as the conversion keeps
+        // only the same number of nibbles as the Max address from console context (PSX : 1FFFFF)
+        viewer.SetAddress(0);
+        viewer.mockEmulatorContext.WriteMemory(0U, MemSize::ThirtyTwoBit, 0x00012345);
```
Your comment says 0x80012345. That's not what you're testing.

Vancleeff committed:
Removed pointless Mask and adjusted tests

Jamiras commented:
Are you still working on these? Three of the four PRs have CI failures that need to be addressed, and the fourth has open comments to resolve.

Vancleeff commented:
Hello @Jamiras !
Yes, I had to finish my second set, it's now on pending review so I will switch back on these PRs !
